src/CommonRuntime/Caching.fs

```diff
@@ -35,16 +36,19 @@ let createInMemoryCache expiration =
 open System.Collections.Concurrent
 
 /// Creates a cache that uses in-memory collection
-let createInMemoryCache expiration = 
+let createInMemoryCache (expiration:TimeSpan) = 
   let dict = new ConcurrentDictionary<_, _>()
   { new ICache<_> with
       member __.Set(key, value) = 
-        dict.[key] <- (value, DateTime.UtcNow)
+        dict.AddOrUpdate(key, (value, DateTime.UtcNow), (fun _ _ -> value, DateTime.UtcNow)) |> ignore
```

Comment:
The AddOrUpdate doesn't give us nothing here, as you are ignoring the old value when calculating the new one